# NHL Goalies Multiple Linear Regression - Statsmodels

In [1]:
# import required modules
import pandas as pd
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.metrics import classification_report, confusion_matrix
import pandas as pd
from sklearn.metrics import roc_auc_score
from sklearn.metrics import precision_recall_fscore_support
import numpy as np
import statsmodels.api as sm

In [2]:
# load dataset
data = pd.read_csv('../../Capstone/Data/NHL_Gpipeline1.csv', header=0)
data.shape

(4340, 26)

In [3]:
data.isna().mean().round(4)*100

Player    0.0
Tm        0.0
Season    0.0
GP        0.0
Age       0.0
GS        0.0
W         0.0
L         0.0
T/O       0.0
GA        0.0
SA        0.0
SV        0.0
GAA       0.0
SO        0.0
GPS       0.0
MIN       0.0
QS        0.0
RBS       0.0
GSAA      0.0
G         0.0
A         0.0
PTS       0.0
PIM       0.0
SV%       0.0
QS%       0.0
GA%       0.0
dtype: float64

In [4]:
data.columns

Index(['Player', 'Tm', 'Season', 'GP', 'Age', 'GS', 'W', 'L', 'T/O', 'GA',
       'SA', 'SV', 'GAA', 'SO', 'GPS', 'MIN', 'QS', 'RBS', 'GSAA', 'G', 'A',
       'PTS', 'PIM', 'SV%', 'QS%', 'GA%'],
      dtype='object')

In [5]:
features=['Age', 'GS',  'GAA',  'QS', 'RBS', 'GSAA', 'G', 'SV%', 'QS%',]
X=data[features]
y=data.GPS

In [6]:
# split X and y into training and testing sets
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=2019)

In [7]:
model = sm.OLS(y, X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                    GPS   R-squared (uncentered):                   0.376
Model:                            OLS   Adj. R-squared (uncentered):              0.375
Method:                 Least Squares   F-statistic:                              289.9
Date:                Mon, 22 Jun 2020   Prob (F-statistic):                        0.00
Time:                        01:39:30   Log-Likelihood:                         -5135.1
No. Observations:                4340   AIC:                                  1.029e+04
Df Residuals:                    4331   BIC:                                  1.035e+04
Df Model:                           9                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Age            0.1858      0.012     15.139      0.000       0.162       0.210
GS            -0.6798      0.049    -13.734      0.000      -0.777      -0.583
GAA            0.1139      0.027      4.192      0.000       0.061       0.167
QS             6.9172      0.329     21.038      0.000       6.273       7.562
RBS            0.0762      0.038      1.993      0.046       0.001       0.151
GSAA           0.2406      0.013     17.956      0.000       0.214       0.267
G              0.1736      0.067      2.591      0.010       0.042       0.305
SV%            0.4099      0.027     15.351      0.000       0.358       0.462
QS%           -6.2715      0.320    -19.611      0.000      -6.899      -5.645
==============================================================================
Omnibus:                       95.333   Durbin-Watson:                   1.924
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               79.957
Skew:                          -0.265   Prob(JB):                     4.34e-18
Kurtosis:                       2.597   Cond. No.                         76.6
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Equation:
GPS = 0.1858(Age) - 0.6798(GS) + 0.1139(GAA) + 6.9172(QS) + 0.0762(RBS) + 0.2406(GSAA) + 0.1736(G) + 0.4099(SV%) - 6.2715(QS%)